packages/enzyme/src/ShallowWrapper.js

```diff
@@ -261,6 +310,25 @@ class ShallowWrapper {
         throw new TypeError('ShallowWrapper can only wrap valid elements');
       }
 
+      if (typeof options.wrappingComponent === 'function') {
+        if (typeof adapter.wrapWithWrappingComponent !== 'function') {
+          throw new TypeError('your adapter does not support `wrappingComponent`. Try upgrading it!');
+        }
+        const { node: wrappedNode, RootFinder } = adapter.wrapWithWrappingComponent(nodes, options);
+        // eslint-disable-next-line no-use-before-define
+        const wrappingComponent = new WrappingComponentWrapper(wrappedNode, this, RootFinder);
+        const wrappingComponentContext = getContextFromWrappingComponent(
+          wrappingComponent, adapter,
+        );
+        options = {
```

Comment:
can any of this logic be moved to `makeOptions`, even if that means "is the root" needs to be passed?

Reply:
Because this is shallow-only logic but `makeOptions` is used in `ShallowWrapper` and `ReactWrapper` I instead opted to pull this logic out into a separate function in `ShallowWrapper.js`.

Reply:
wait - why is this shallow-only? `mount` should be able to have a wrapping component as well.

Reply:
`mount` can have a wrapping component! I just meant all the code that I moved out of the constructor of `ShallowWrapper` into a separate function is only for shallow. That code:

* Does the recursive `dive()`ing
* Pulls the accumulated context off of the `<RootFinder />`
* Transforms `options.context` to extend it with the accumulated context

None of that should happen when mounting because we just _actually_ render the wrapping component and let react just do its thing w/r/t the context.